README.md
# BBTAG-Improvement-Mod
Additional features and tweaks for BlazBlue Cross Tag Battle

The goal is to extend the game with community-created content, provide additional graphical options, and improve performance for players with computer specs below the minimal system requirements.

Join the [BB Improvement Mod Discord Server](https://discord.gg/j2mCX9s) to discuss about the development and the game itself!

## Legal
```
BBTAG Improvement Mod is NOT associated with Arc System Works or any of its partners / affiliates.
BBTAG Improvement Mod is NOT intended for malicious use.
BBTAG Improvement Mod is NOT intended to give players unfair advantages in online matches.
BBTAG Improvement Mod is NOT intended to unlock unreleased / unpurchased contents of the game.
BBTAG Improvement Mod should only be used on the official version that you legally purchased and own.

Use BBTAG Improvement Mod at your own risk. 
I, KoviDomi, am not responsible for what happens while using BBTAG Improvement Mod. You take full reponsibility for any outcome that happens to you while using this application.
```

## Compiling and usage
BBTAG Improvement Mod is coded using Visual Studio 2017 (toolset v141). <br>
To compile, you should only need to load the sln file and compile as-is. No changes should be needed to the solution or source.<br>
Copy the compiled binary, settings.ini, and palettes.ini file from the bin/ folder into the game's root folder.

## Requirements:
- Visual Studio 2017 (toolset v141) (Windows SDK 10.0.17763.0)
- Detours 1.5
- DirectX SDK
- Steamworks SDK (v1.42)

## Quick file overview
- **info.h** <br>
Defines for version numbering and outside links
- **logging.h** <br>
Self explanatory
- **interfaces.h** <br>
Access to all the wrappers and hooked game variables
- **D3D9ExWrapper/** <br>
Straightforward D3D wrapping
- **Game/** <br>
Player, ingame characters, meters information
- **PaletteManager/** <br>
Loading and in-game management of custom palettes
- **SteamApiWrapper/** <br>
Straightforward Steam API wrapping
- **Web/** <br>
HTTP request functions for update checking, palette downloading, fetching donator names
- **WindowManager/** <br>
Implementation of the mod's windows using the ImGui library
- **Hooks/hooks_bbtag** <br>
Implements game functions overriding
- **Hooks/hooks_detours** <br>
Implements imported functions overriding using the Detours library
- **Hooks/hooks_steamApiWrapper** <br>
Implements Steam API overriding


## Support
If you feel like throwing some dollars my way feel free to, it enables me spending more time working on the mod.

[![Donate](https://img.shields.io/badge/Donate-PayPal-green.svg)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=C8EDR78DJ8PU2)

[![paypal](https://www.paypalobjects.com/en_US/i/btn/btn_donateCC_LG.gif)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=C8EDR78DJ8PU2)

## Thanks to the people who have helped the mod along the way:
* Neptune
* Rouzel
* Labryz
* Dormin
* NeoStrayCat
* Everybody in the BlazBlue PC community that has given support or feedback of any kind!

## Extra thanks to:
Atom0s for their DirectX9.0 Hooking article<br>
Durante for their dsfix source


CHANGELOG.txt
11-12-2019 -- version 1.09
===============================
- Updated for official patch v2.2
- Added dinput dll chaining to settings.ini to use other dinput8.dll wrappers together with the mod
- Buttons that open windows now toggle them instead of just opening
- Added button to open the mod's download page upon new update on the update notification window
- Reworked donators window
- Slight visual changes
- Fixed crash upon assigning custom palettes to slots 17-20 in palettes.ini

24-11-2019 -- version 1.08
===============================
- Updated for official patch v2.0
- Updated "palettes.ini" file for new characters and extra palette slots
- Enabled palette editor in versus mode
- Hitbox overlay now can be toggled for each characters separately
- Added frame freezing option to hitbox overlay
- Added frame stepping function to hitbox overlay
- Added tooltip to show the character names on the custom palette selection buttons
- The online palettes archive is now only downloaded if there has been a change
- Fixed crash due to non-existent character index in manually edited impl file headers

22-05-2019 -- version 1.07
===============================
- Updated for official patch v1.50
- Added hitbox overlay in training mode
    * Calculations are not pixel perfect, expect slight margin of error
    * Personas upon disappearing may leave their hitboxes behind in weird ways
- Added gradient generator to the palette editor
- Added indexing mode to the palette editor
- "Drag and drop" of the color boxes now works on the palette editor
- Palettes are now previewed when hovered over their selections
- Fixed being able to manually resize the game's window using the cursor in windowed mode

07-08-2018 -- version 1.06
===============================
- Added character specific meters to the custom HUD for Akihiko, Labrys, and Nine
- Custom HUD now shows astral availability with golden borders around the gauge window
- The game does not need to be restarted after making changes to the "palettes.ini" file
- Fixed palette editor button's text from "not in tutorial mode" to "not in training mode"

29-07-2018 -- version 1.05
===============================
- Fixed crash when fighting against the final boss in Story Mode

28-07-2018 -- version 1.04
===============================
- Added ingame palette editor
- Added "palettes.ini" file for assigning custom palettes to the ingame palette slots
- Added additional option for ForcePromptButtonLayout to keep the button prompts blinking
- Added donators window
- Added button links for the mod's GitHub, and donation pages
- Adjusted custom HUD window sizes to take up less space on extreme low resolutions
- The mod's main window won't disappear if you click on it within 10 seconds after the title screen
- Fixed the custom HUD not being updated while the game's window is not focused
- Fixed mistakenly calling the Steam Controller as PS4 controller all over the place

21-06-2018 -- version 1.03
===============================
- Added character specific meters to the custom HUD for Aegis, Chie, Makoto, Naoto, Platinum, Rachel, and Yukiko
- Added CustomHUDScale option
- Added PromptButtonLayout option that prevents the game from blinking the button prompts between PS4 and PC layouts
- Added a tooltip to the "Reset custom HUD positions" button to avoid confusion about its purpose
- The game's window is now always positioned to the middle of the screen when DisplayMode is set to Windowed(1) in the settings.ini
- The mod's main window now automatically disappears within 10 seconds after the title screen
- Fixed custom HUD HP bars not switching immediately after characters die
- Fixed alt tabbing in fullscreen causing all custom HUD elements being thrown into the top left corner

16-06-2018 -- version 1.02
===============================
- Added Custom HUD
- Added ToggleCustomHUDButton option
- Added ForceCustomHUD option
- Added DisplayMode option
- Added FullscreenHz option
- Added "Reset custom HUD positions" button
- Removed palette folders being created in BBTAG_IM folder

07-06-2018 -- version 1.01
===============================
- Added togglable HUD option (temporary solution for performance problems during matches)
- Fixed gamewindow getting keyboard inputs while not in focus

06-06-2018 -- version 1.00
===============================
- Initial release

src/charobj.h
#pragma once
#include <stdint.h>

#define TOTAL_CHAR_INDEXES 41

extern char* charnames[TOTAL_CHAR_INDEXES];
extern wchar_t* wcharnames[TOTAL_CHAR_INDEXES];

enum CharIndex
{
	CharIndex_Ragna		= 0,
	CharIndex_Jin		= 1,
	CharIndex_Noel		= 2,
	CharIndex_Rachel	= 3,
	CharIndex_Tager		= 4,
	CharIndex_Hakumen	= 5,
	CharIndex_Nu		= 6,
	CharIndex_Hazama	= 7,
	CharIndex_Makoto	= 8,
	CharIndex_Platinum	= 9,
	CharIndex_Izayoi	= 10,
	CharIndex_Azrael	= 11,
	CharIndex_Nine		= 12,
	CharIndex_Es		= 13,
	CharIndex_Mai		= 14,
	CharIndex_Jubei		= 15,
	CharIndex_Yu		= 16,
	CharIndex_Yosuke	= 17,
	CharIndex_Chie		= 18,
	CharIndex_Yukiko	= 19,
	CharIndex_Kanji		= 20,
	CharIndex_Naoto		= 21,
	CharIndex_Mitsuru	= 22,
	CharIndex_Akihiko	= 23,
	CharIndex_Aegis		= 24,
	CharIndex_Labrys	= 25,
	CharIndex_Hyde		= 26,
	CharIndex_Linne		= 27,
	CharIndex_Waldstein	= 28,
	CharIndex_Carmine	= 29,
	CharIndex_Orie		= 30,
	CharIndex_Gordeau	= 31,
	CharIndex_Merkava	= 32,
	CharIndex_Vatista	= 33,
	CharIndex_Yuzuriha	= 34,
	CharIndex_Mika		= 35,
	CharIndex_Ruby		= 36,
	CharIndex_Weiss		= 37,
	CharIndex_Blake		= 38,
	CharIndex_Yang		= 39,
	CharIndex_Boss		= 40,
};

class CharObj
{
public:
	char pad_0000[52]; //0x0000
	uint32_t char_index; //0x0034
	char pad_0038[368]; //0x0038
	uint32_t chie_charge_lvl; //0x01A8
	char pad_01AC[2296]; //0x01AC
	uint32_t cur_hp; //0x0AA4
	uint32_t max_hp; //0x0AA8
	char pad_0AAC[7504]; //0x0AAC
	uint32_t is_char_active2; //0x27FC
	char pad_2800[4]; //0x2800
	uint32_t recoverable_hp; //0x2804
	uint32_t is_char_active; //0x2808
	char pad_280C[116580]; //0x280C
	uint32_t naoto_is_enemy_marked; //0x1EF70
	char pad_1EF74[8540]; //0x1EF74
	uint32_t unique_meter_max_val; //0x210D0
	uint32_t unique_meter_cur_val; //0x210D4
	char pad_210D8[40]; //0x210D8
}; //Size: 0x21100